jupyterhub/spawner.py

```diff
@@ -247,8 +250,32 @@ def name(self):
             </select>
 
         The data from this form submission will be passed on to your spawner in `self.user_options`
+
+        Instead of a form snippet string, this could also be a callable that takes as one
+        parameter the current spawner instance and returns a string. The callable will
+        be called asynchronously if it returns a future, rather than a str. Note that
+        the interface of the spawner class is not deemed stable across versions,
+        so using this functionality might cause your JupyterHub upgrades to break.
     """).tag(config=True)
 
+    @gen.coroutine
+    def get_options_form(self):
+        """Get the options form
+
+        Returns:
+          (Future(str)): the content of the options form presented to the user
+          prior to starting a Spawner.
+
+        .. versionchanged:: 0.8.2
```

Comment:
This can be `.. versionadded: 0.9`

Reply:
Great, thanks for the feedback! Updated.